SourceX/qol.cpp

```diff
@@ -237,4 +237,156 @@ void AutoGoldPickup(int pnum)
+bool isLabelHighlighted = false;
+int labelCenterOffsets[ITEMTYPES];
+
+void AltPressed(bool pressed)
+{
+	altPressed = pressed;
+}
+
+bool IsGeneratingLabels()
+{
+	return generatingLabelsState == 1;
+}
+
+bool IsLabelHighlighted()
+{
+	return isLabelHighlighted;
+}
+
+void UpdateLabelOffsets(CelOutputBuffer out, BYTE *dst, int width)
+{
+	int xval = (dst - out.begin()) % out.pitch();
+	if (xval < drawMinX)
+		drawMinX = xval;
+	xval += width;
+	if (xval > drawMaxX)
+		drawMaxX = xval;
+}
+
+void GenerateLabelOffsets(CelOutputBuffer out)
+{
+	if (generatingLabelsState == 2)
+		return;
+	generatingLabelsState = 1;
+	int itemTypes = gbIsHellfire ? ITEMTYPES : 35;
+	for (int i = 0; i < itemTypes; i++) {
+		drawMinX = gnScreenWidth;
+		drawMaxX = 0;
+		CelClippedDrawTo(out, out.pitch() / 2 - 16, 351, itemanims[i], ItemAnimLs[i], 96);
+		labelCenterOffsets[i] = drawMinX - out.pitch() / 2 + (drawMaxX - drawMinX) / 2;
+	}
+	generatingLabelsState = 2;
+}
+
+void AddItemToDrawQueue(int x, int y, int id)
+{
+	if (sgOptions.Gameplay.bAltHighlight == 0 || (sgOptions.Gameplay.bAltHighlight == 1 && !altPressed) || (sgOptions.Gameplay.bAltHighlight == 2 && altPressed))
+		return;
+	ItemStruct *it = &items[id];
+
+	char textOnGround[64];
+	if (it->_itype == ITYPE_GOLD)
+		sprintf(textOnGround, "%i gold", it->_ivalue);
+	else
+		sprintf(textOnGround, "%s", it->_iIdentified ? it->_iIName : it->_iName);
+
+	int nameWidth = GetTextWidth((char *)textOnGround);
+	x += labelCenterOffsets[ItemCAnimTbl[it->_iCurs]];
+	y -= TILE_HEIGHT;
+	if (!zoomflag) {
+		x <<= 1;
+		y <<= 1;
+	}
+	x -= nameWidth >> 1;
+	text_color clr = COL_WHITE;
+	if (it->_iMagical == ITEM_QUALITY_MAGIC)
+		clr = COL_BLUE;
+	if (it->_iMagical == ITEM_QUALITY_UNIQUE)
+		clr = COL_GOLD;
+	drawQ.push_back(drawingQueue(id, it->_ix, it->_iy, x, y, nameWidth, 13, clr, textOnGround));
+}
+
+void ShowItemNameLabels(CelOutputBuffer out)
+{
+	isLabelHighlighted = false;
+	const int borderX = 5;
+	for (unsigned int i = 0; i < drawQ.size(); ++i) {
+		std::map<int, bool> backtrace;
+
+		bool canShow;
+		do {
+			canShow = true;
+			for (unsigned int j = 0; j < i; ++j) {
+				if (abs(drawQ[j].y - drawQ[i].y) < drawQ[i].height + 2) {
+					int newpos = drawQ[j].x;
+					if (drawQ[j].x >= drawQ[i].x && drawQ[j].x - drawQ[i].x < drawQ[i].width + borderX) {
+						newpos -= drawQ[i].width + borderX;
+						if (backtrace.find(newpos) != backtrace.end())
+							newpos = drawQ[j].x + drawQ[j].width + borderX;
+					} else if (drawQ[j].x < drawQ[i].x && drawQ[i].x - drawQ[j].x < drawQ[j].width + borderX) {
+						newpos += drawQ[j].width + borderX;
+						if (backtrace.find(newpos) != backtrace.end())
+							newpos = drawQ[j].x - drawQ[i].width - borderX;
+					} else
+						continue;
+					canShow = false;
+					drawQ[i].x = newpos;
+					backtrace[newpos] = true;
+				}
+			}
+		} while (!canShow);
+	}
+
+	for (unsigned int i = 0; i < drawQ.size(); ++i) {
+		drawingQueue t = drawQ[i];

```

Comment:
Doesn't look like you need to copy it out, or that you use the index;

```suggestion
	for (const drawingQueue &t : drawQ) {
```